README.md
# isp README

isp extension used to download .hex to stm32 hardware.

## Features

just download .hex.

## Extension Settings

This extension contributes the following settings:

* `isp.hex`: .hex file path
* `isp.com`: serial port
* `isp.baud`: serial baud rate


CHANGELOG.md
# Change Log

All notable changes to the "isp" extension will be documented in this file.

### 0.0.1
get all start.
### 0.0.9
read workspace configure and accept relative path for hex.
### 0.0.10
fixed relative path reading bug.
### 0.0.11
update serialport lib.
### 0.0.12
update serialport lib.
### 0.0.12
dependend on `ms-vscode.vscode-serial-monitor` for serialport lib.


src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import Stm32Isp from './stm32isp';

let localize = {} as any;
try {
  try {
    localize = eval(`require("../package.nls.${JSON.parse(process.env.VSCODE_NLS_CONFIG as string).locale.toLowerCase()}.json")`);
  } catch {
    localize = eval(`require("../package.nls.json")`);
  }
} catch { }

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export async function activate(context: vscode.ExtensionContext) {
	const statusbar = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left);
	statusbar.tooltip = localize?.['extension.tool.tips'] || 'ISP download hex to board';
	statusbar.command = 'isp.execute';
  const title = localize?.['extension.tool.title'] || 'Download HEX';
	statusbar.text = '$(sort-precedence) ' + title;
	statusbar.show();

	const channel = vscode.window.createOutputChannel('isp');

	let disposable = vscode.commands.registerCommand('isp.execute', async () => {
		channel.show();
		channel.clear();
		channel.appendLine('start downloading...');
		let isp!: Stm32Isp;
		try {
			isp = new Stm32Isp(channel);
			await isp.run();
			channel.appendLine('all done');
		} catch (e) {
			channel.appendLine((e as Error).stack || e as any);
		} finally {
			isp?.close();
		}
		// vscode.window.showInformationMessage('hex path is ' + cfg.get('hex'));
	});

	context.subscriptions.push(disposable);
}

export function deactivate() { }


webpack.config.js
//@ts-check

'use strict';

const path = require('path');
// const copyPlugin = require("copy-webpack-plugin");

//@ts-check
/** @typedef {import('webpack').Configuration} WebpackConfig **/

/** @type WebpackConfig */
const extensionConfig = {
  target: 'node', // vscode extensions run in a Node.js-context 📖 -> https://webpack.js.org/configuration/node/
  mode: 'none', // this leaves the source code as close as possible to the original (when packaging we set this to 'production')

  entry: './src/extension.ts', // the entry point of this extension, 📖 -> https://webpack.js.org/configuration/entry-context/
  output: {
    // the bundle is stored in the 'dist' folder (check package.json), 📖 -> https://webpack.js.org/configuration/output/
    path: path.resolve(__dirname, 'dist'),
    filename: 'extension.js',
    libraryTarget: 'commonjs2',
    assetModuleFilename: '[name][ext]'
  },
  externals: {
    serialport: 'serialport',
    vscode: 'commonjs vscode' // the vscode-module is created on-the-fly and must be excluded. Add other modules that cannot be webpack'ed, 📖 -> https://webpack.js.org/configuration/externals/
    // modules added here also need to be added in the .vscodeignore file
  },
  watchOptions: {
    ignored: [path.posix.resolve(__dirname, './dist')]
  },
  resolve: {
    // support reading TypeScript and JavaScript files, 📖 -> https://github.com/TypeStrong/ts-loader
    extensions: ['.ts', '.js']
  },
  module: {
    rules: [
      {
        test: /\.ts$/,
        exclude: /node_modules/,
        use: [
          {
            loader: 'ts-loader'
          }
        ]
      }
    ]
  },
  devtool: 'nosources-source-map',
  infrastructureLogging: {
    level: "log", // enables logging required for problem matchers
  },
  // plugins: [
  //   new copyPlugin({
  //     patterns: [
  //       'package.nls.*'
  //     ]
  //   })
  // ]
};
module.exports = [extensionConfig];

src/stm32isp/hex.ts
/* eslint-disable no-throw-literal */
import * as fs from 'fs';

function nextLine(buf: string, from: number) {
  for (let i = from; i < buf.length; i++) {
    if (buf[i] === ':') {
      const line = {
        index: i,
        size: parseInt('0x' + buf.substring(i + 1, i + 3)),
        address: parseInt('0x' + buf.substring(i + 3, i + 7)),
        type: parseInt('0x' + buf.substring(i + 7, i + 9)),
        data: [] as number[]
      };
      for (let j = 0; j < line.size; j++) {
        line.data.push(parseInt('0x' + buf.substring(i + 9 + 2 * j, i + 9 + 2 * j + 2)));
      }
      return line;
    }
  }
  return null;
}

function getLine(buf: string, type: number, from: number) {
  while (from < buf.length) {
    const line = nextLine(buf, from);
    if (!line) {
      throw 'can not get line ' + type + ' from ' + from;
    }
    if (line.type === type) {
      return line;
    }
    from += line.size + 10;
  }
  throw 'can not get line ' + type + ' from ' + from;
}

function hex2bin(hpath: string) {
  const buf = fs.readFileSync(hpath, { encoding: 'ascii' });
  let line = getLine(buf, 4, 0);
  let address = (line.data[0] << 8) + line.data[1];
  line = getLine(buf, 0, 12);
  address = (address << 16) + line.address;
  if (address !== 0x08000000) {
    throw 'base address is not 0x08000000';
  }
  const data = [];
  for (let i = 0; i < buf.length;) {
    try {
      line = getLine(buf, 0, i);
      data.push(...line.data);
      i = line.index + line.size + 10;
    } catch {
      break;
    }
  }
  return data;
}

export default hex2bin;


src/stm32isp/index.ts
/* eslint-disable no-throw-literal */
import * as fs from 'fs';
import * as path from 'path';
import * as vscode from 'vscode';
import { SerialPort, SerialPortOpenOptions } from 'serialport';

import hex2bin from './hex';

export default class Stm32Isp {
  private port!: SerialPort;
  private lastData = [] as number[];
  private dataResolve: ((data: number[]) => void) | null = null;
  private hexData: number[] = [];
  private channel!: vscode.OutputChannel;
  private serialPortCtor!: new (..._args: any[]) => SerialPort;

  constructor(ch: vscode.OutputChannel) {
    this.loadLib();
    this.channel = ch;
    const cfg = vscode.workspace.getConfiguration('isp');
    let hexPath = cfg.get('hex') as string;
    const comPort = cfg.get('com') as string || 'COM3';
    const baudRate = cfg.get('baud') as number || 115200;

    if (!hexPath) {
      throw 'can not get hex file path';
    }
    if (!path.isAbsolute(hexPath) && vscode.workspace.workspaceFolders) {
      hexPath = path.join(vscode.workspace.workspaceFolders[0].uri.fsPath, hexPath);
    }
    if (!fs.existsSync(hexPath)) {
      throw `hex ${hexPath} not exists`; vscode.workspace.rootPath;
    }
    this.hexData = hex2bin(hexPath);

    this.port = new this.serialPortCtor({
      path: comPort,
      baudRate,
      parity: 'even',
      dataBits: 8,
      stopBits: 1,
      autoOpen: false
    } as SerialPortOpenOptions<any>);

    this.port.on('data', (data: number[]) => {
      this.dataResolve?.(data);
    });
  }

  private loadLib() {
    try {
      const serialMonitorExt = vscode.extensions.all.find(v => v.id === 'ms-vscode.vscode-serial-monitor');
      const pp = path.join((serialMonitorExt?.extensionPath as string), './dist/node_modules', 'serialport');
      this.serialPortCtor = eval(`require("${pp.replace(/\\/g, '\\\\')}").SerialPort`);
    } catch {
      throw 'can not load SerialPort library!';
    }
  }

  private waitAck(timeout = 3000) {
    return new Promise(res => {
      const timer = setTimeout(() => {
        this.dataResolve = null;
        res(false);
      }, timeout);
      this.dataResolve = (data) => {
        clearTimeout(timer);
        this.lastData = [...data];
        res(data[0] === 0x79);
      };
    });
  }

  private open() {
    return new Promise(res => {
      this.port.open((err) => {
        res(!err);
      });
    });
  }

  private checksum(data: number[]) {
    let cs = 0;
    for (let i = 0; i < data.length; i++) { cs ^= data[i]; }
    return cs;
  }

  private async sync() {
    this.channel.appendLine('get syncing...');
    let retry = 10;
    while (true) {
      this.port.write([0x7F]);
      if (!await this.waitAck(1000) && --retry > 0) {
        this.channel.appendLine('please reset board, try again');
        continue;
      } else {
        if (retry < 1) {
          throw 'too many times retry, exit';
        }
        break;
      }
    }
  }

  private async getVersion() {
    this.port.write([0x00]);
    this.port.write([0xff]);
    if (!await this.waitAck()) {
      throw 'get cmd timeout';
    }
    if (this.lastData[1] === 11) {
      this.channel.appendLine(`version: ${this.lastData[2] >> 4}.${this.lastData[2] & 0x0F}`);
    } else {
      throw 'get cmd error: ' + this.lastData.join(',');
    }
  }

  private async getID() {
    this.port.write([0x02]);
    this.port.write([0xfd]);
    if (!await this.waitAck()) {
      throw 'get id timeout';
    }
    if (this.lastData[1] === 1) {
      this.channel.appendLine(`pid: 0x${this.lastData[2].toString(16)}${this.lastData[3].toString(16)}`);
    } else {
      throw 'get id error: ' + this.lastData.join(',');
    }
  }

  private async erase() {
    this.channel.appendLine('erase all');
    this.port.write([0x43]);
    this.port.write([0xbc]);
    if (!await this.waitAck()) {
      throw 'erase timeout';
    }
    this.port.write([0xff]);
    this.port.write([0x00]);
    if (!await this.waitAck()) {
      throw 'erase timeout';
    }
  }

  private async write() {
    this.channel.appendLine('writing...');
    let addr = 0x08000000;
    let temp = [];
    let len = 256;
    for (let i = 0; i < this.hexData.length;) {
      temp[0] = ((addr >> 24) & 0xff);
      temp[1] = ((addr >> 16) & 0xff);
      temp[2] = ((addr >> 8) & 0xff);
      temp[3] = ((addr) & 0xff);

      this.port.write([0x31]);
      this.port.write([0xce]);
      if (!await this.waitAck()) {
        throw 'write data error';
      }
      this.port.write(temp);
      this.port.write([this.checksum(temp)]);
      if (!await this.waitAck()) {
        throw 'write data error';
      }
      len = Math.min(256, this.hexData.length - i);
      //下面发送数据
      this.port.write([len - 1]);
      this.port.write(this.hexData.slice(i, i + len));
      this.port.write([(len - 1) ^ this.checksum(this.hexData.slice(i, i + len))]);
      if (!await this.waitAck()) {
        throw 'write data error';
      }
      this.port.flush();
      addr += len;
      i += len;
    }
  }

  private async varify() {
    this.channel.appendLine('varifing');
    let addr = 0x08000000;
    let temp = [];
    let len = 256;
    for (let i = 0; i < this.hexData.length;) {
      temp[0] = ((addr >> 24) & 0xff);
      temp[1] = ((addr >> 16) & 0xff);
      temp[2] = ((addr >> 8) & 0xff);
      temp[3] = ((addr) & 0xff);

      this.port.write([0x11]);
      this.port.write([0xEE]);
      if (!await this.waitAck()) {
        throw 'read data error';
      }
      this.port.write(temp);
      this.port.write([this.checksum(temp)]);
      if (!await this.waitAck()) {
        throw 'read data error';
      }
      len = Math.min(256, this.hexData.length - i);
      //下面发送数据
      this.port.write([len - 1]);
      this.port.write([~(len - 1)]);
      if (!await this.waitAck()) {
        throw 'read data error';
      }
      for (let j = 0; j < len; j++) {
        if (this.hexData[i + j] !== this.lastData[1 + j]) {
          throw `varify ${i} ${this.lastData.length}: ${this.lastData.join(',')}`;
        }
      }
      addr += len;
      i += len;
    }
  }

  private async execute() {
    let temp = [];
    let addr = 0x08000000;
    temp[0] = ((addr >> 24) & 0xff);
    temp[1] = ((addr >> 16) & 0xff);
    temp[2] = ((addr >> 8) & 0xff);
    temp[3] = ((addr) & 0xff);

    this.port.write([0x21]);
    this.port.write([0xDE]);
    if (!await this.waitAck()) {
      throw 'execute error';
    }
    this.port.write(temp);
    this.port.write([this.checksum(temp)]);
    if (!await this.waitAck()) {
      throw 'execute error';
    }
    this.channel.appendLine('execute app from 0x08000000');
  }

  async run() {
    if (!await this.open()) {
      throw 'can not open serial port';
    }
    this.port.flush();
    await this.sync();
    await this.getVersion();
    await this.getID();
    await this.sleep(100);
    await this.erase();
    await this.sleep(100);
    await this.write();
    await this.sleep(100);
    await this.varify();
    await this.sleep(100);
    await this.execute();
  }

  close() {
    if (this.port.isOpen) {
      this.port.close();
    }
  }

  private sleep(ms: number) {
    return new Promise(res => {
      setTimeout(res, ms);
    });
  }
}
